File: pixsim7/backend/main/domain/narrative/ecs_helpers.py
"""
Narrative Runtime ECS Helpers (canonical GameObject store).

Narrative state lives as a ``{"type": "narrative", "data": {...}}`` entry in
the canonical npc GameObject's ``components`` array, persisted at
``session.flags["gameObjects"]["objects"]["npc:<id>"].components``.

The legacy location ``session.flags.npcs[<key>].components.narrative`` is no
longer written or read (this module was its sole writer; no users / no data per
plan ``backend-canonical-gameobject-adoption``). Reads and writes route through
``services.game.game_object_store``.
"""

from __future__ import annotations
from typing import Any, Dict, List, Optional
import time

from pixsim7.backend.main.domain.game.core.models import GameSession
from pixsim7.backend.main.domain.narrative.schema import (
    NarrativeRuntimeState,
    NarrativeProgramId,
    NodeId,
    StackFrame,
    HistoryEntry,
    ErrorState,
)
from pixsim7.backend.main.services.game.game_object_store import (
    get_npc_component,
    remove_npc_component,
    set_npc_component,
)


# ============================================================================
# Canonical npc-component Access
# ============================================================================

NARRATIVE_COMPONENT_TYPE = "narrative"


def _empty_narrative_state() -> NarrativeRuntimeState:
    return NarrativeRuntimeState(
        active_program_id=None,
        active_node_id=None,
        stack=[],
        history=[],
        variables={},
        last_step_at=None,
        paused=False,
        error=None,
    )


def get_narrative_state(session: GameSession, npc_id: int) -> NarrativeRuntimeState:
    """Get narrative runtime state for an NPC, or a fresh/empty state."""
    comp = get_npc_component(
        session.flags, session.world_id, npc_id, NARRATIVE_COMPONENT_TYPE
    )
    if comp is not None and isinstance(comp.get("data"), dict):
        return NarrativeRuntimeState(**comp["data"])
    return _empty_narrative_state()


def set_narrative_state(
    session: GameSession, npc_id: int, state: NarrativeRuntimeState
) -> None:
    """Write narrative state to the canonical npc's ``narrative`` component."""
    set_npc_component(
        session.flags,
        session.world_id,
        npc_id,
        NARRATIVE_COMPONENT_TYPE,
        state.model_dump(mode="json"),
    )


def clear_narrative_state(session: GameSession, npc_id: int) -> None:
    """Remove the narrative component from the canonical npc (no-op when absent)."""
    remove_npc_component(
        session.flags, session.world_id, npc_id, NARRATIVE_COMPONENT_TYPE
    )


# ============================================================================
# Program Lifecycle Helpers
# ============================================================================

def start_program(
    session: GameSession,
    npc_id: int,
    program_id: NarrativeProgramId,
    entry_node_id: NodeId,
    initial_variables: Optional[Dict[str, Any]] = None
) -> NarrativeRuntimeState:
    """
    Start a new narrative program for an NPC.

    If a program is already active, it will be pushed to the stack (nested program).

    Args:
        session: Game session
        npc_id: NPC ID
        program_id: Program ID to start
        entry_node_id: Entry node ID
        initial_variables: Initial program variables

    Returns:
        Updated narrative state
    """
    state = get_narrative_state(session, npc_id)

    # If there's an active program, push it to the stack
    if state.active_program_id and state.active_node_id:
        state.stack.append(
            StackFrame(
                program_id=state.active_program_id,
                node_id=state.active_node_id,
                pushed_at=int(time.time())
            )
        )

    # Set new active program
    state.active_program_id = program_id
    state.active_node_id = entry_node_id
    state.variables = initial_variables or {}
    state.last_step_at = int(time.time())
    state.paused = False
    state.error = None

    # Add to history
    state.history.append(
        HistoryEntry(
            program_id=program_id,
            node_id=entry_node_id,
            timestamp=int(time.time())
        )
    )

    # Save state
    set_narrative_state(session, npc_id, state)

    return state


def finish_program(
    session: GameSession,
    npc_id: int
) -> Optional[NarrativeRuntimeState]:
    """
    Finish the currently active program for an NPC.

    If there are programs on the stack, pops and resumes the previous one.
    Otherwise, clears the narrative state.

    Args:
        session: Game session
        npc_id: NPC ID

    Returns:
        Updated narrative state, or None if no programs remain
    """
    state = get_narrative_state(session, npc_id)

    if not state.active_program_id:
        # No active program
        return None

    # Pop from stack if available
    if state.stack:
        frame = state.stack.pop()
        state.active_program_id = frame.program_id
        state.active_node_id = frame.node_id
        state.last_step_at = int(time.time())

        # Add resume to history
        state.history.append(
            HistoryEntry(
                program_id=frame.program_id,
                node_id=frame.node_id,
                timestamp=int(time.time())
            )
        )

        set_narrative_state(session, npc_id, state)
        return state
    else:
        # No more programs, clear state
        clear_narrative_state(session, npc_id)
        return None


def advance_to_node(
    session: GameSession,
    npc_id: int,
    node_id: NodeId,
    choice_id: Optional[str] = None,
    edge_id: Optional[str] = None
) -> NarrativeRuntimeState:
    """
    Advance to a new node in the current program.

    Args:
        session: Game session
        npc_id: NPC ID
        node_id: Target node ID
        choice_id: Choice ID if advancing via choice
        edge_id: Edge ID if advancing via edge

    Returns:
        Updated narrative state
    """
    state = get_narrative_state(session, npc_id)

    if not state.active_program_id:
        raise ValueError("No active program to advance")

    # Update active node
    state.active_node_id = node_id
    state.last_step_at = int(time.time())

    # Add to history
    state.history.append(
        HistoryEntry(
            program_id=state.active_program_id,
            node_id=node_id,
            timestamp=int(time.time()),
            choice_id=choice_id,
            edge_id=edge_id
        )
    )

    # Save state
    set_narrative_state(session, npc_id, state)

    return state


def pause_program(
    session: GameSession,
    npc_id: int
) -> NarrativeRuntimeState:
    """
    Pause the currently active program.

    Args:
        session: Game session
        npc_id: NPC ID

    Returns:
        Updated narrative state
    """
    state = get_narrative_state(session, npc_id)

    if not state.active_program_id:
        raise ValueError("No active program to pause")

    state.paused = True
    set_narrative_state(session, npc_id, state)

    return state


def resume_program(
    session: GameSession,
    npc_id: int
) -> NarrativeRuntimeState:
    """
    Resume a paused program.

    Args:
        session: Game session
        npc_id: NPC ID

    Returns:
        Updated narrative state
    """
    state = get_narrative_state(session, npc_id)

    if not state.active_program_id:
        raise ValueError("No active program to resume")

    state.paused = False
    set_narrative_state(session, npc_id, state)

    return state


def set_error(
    session: GameSession,
    npc_id: int,
    error_message: str,
    node_id: NodeId
) -> NarrativeRuntimeState:
    """
    Set error state for the current program.

    Args:
        session: Game session
        npc_id: NPC ID
        error_message: Error message
        node_id: Node where error occurred

    Returns:
        Updated narrative state
    """
    state = get_narrative_state(session, npc_id)

    state.error = ErrorState(
        message=error_message,
        node_id=node_id,
        timestamp=int(time.time())
    )

    set_narrative_state(session, npc_id, state)

    return state


def clear_error(
    session: GameSession,
    npc_id: int
) -> NarrativeRuntimeState:
    """
    Clear error state.

    Args:
        session: Game session
        npc_id: NPC ID

    Returns:
        Updated narrative state
    """
    state = get_narrative_state(session, npc_id)

    state.error = None
    set_narrative_state(session, npc_id, state)

    return state


# ============================================================================
# Query Helpers
# ============================================================================

def is_program_active(
    session: GameSession,
    npc_id: int,
    program_id: Optional[NarrativeProgramId] = None
) -> bool:
    """
    Check if a program is currently active.

    Args:
        session: Game session
        npc_id: NPC ID
        program_id: Optional specific program ID to check

    Returns:
        True if program is active
    """
    state = get_narrative_state(session, npc_id)

    if not state.active_program_id:
        return False

    if program_id:
        return state.active_program_id == program_id

    return True


def get_program_variable(
    session: GameSession,
    npc_id: int,
    variable_name: str,
    default: Any = None
) -> Any:
    """
    Get a program variable value.

    Args:
        session: Game session
        npc_id: NPC ID
        variable_name: Variable name
        default: Default value if not found

    Returns:
        Variable value or default
    """
    state = get_narrative_state(session, npc_id)
    return state.variables.get(variable_name, default)


def set_program_variable(
    session: GameSession,
    npc_id: int,
    variable_name: str,
    value: Any
) -> NarrativeRuntimeState:
    """
    Set a program variable value.

    Args:
        session: Game session
        npc_id: NPC ID
        variable_name: Variable name
        value: Variable value

    Returns:
        Updated narrative state
    """
    state = get_narrative_state(session, npc_id)
    state.variables[variable_name] = value
    set_narrative_state(session, npc_id, state)
    return state


def has_visited_node(
    session: GameSession,
    npc_id: int,
    program_id: NarrativeProgramId,
    node_id: NodeId
) -> bool:
    """
    Check if a node has been visited in the history.

    Args:
        session: Game session
        npc_id: NPC ID
        program_id: Program ID
        node_id: Node ID

    Returns:
        True if node has been visited
    """
    state = get_narrative_state(session, npc_id)

    for entry in state.history:
        if entry.program_id == program_id and entry.node_id == node_id:
            return True

    return False


def get_stack_depth(
    session: GameSession,
    npc_id: int
) -> int:
    """
    Get the current call stack depth.

    Args:
        session: Game session
        npc_id: NPC ID

    Returns:
        Stack depth (0 if no active program)
    """
    state = get_narrative_state(session, npc_id)
    return len(state.stack)

File: packages/game/engine/src/narrative/ecsHelpers.ts
/**
 * Narrative Runtime ECS Helpers
 *
 * Provides helper functions for managing narrative runtime state in the ECS component system.
 * Narrative state is stored at: session.flags.npcs["npc:<id>"].components.narrative
 *
 * TypeScript equivalent of: pixsim7/backend/main/domain/narrative/ecs_helpers.py
 */

import type {
  NarrativeRuntimeState,
  NarrativeProgramId,
  NodeId,
} from '@pixsim7/shared.types';
import { Ref } from '@pixsim7/ref-core';

/**
 * Session flags structure (subset needed for narrative)
 */
interface SessionFlags {
  npcs?: Record<string, {
    components?: Record<string, any>;
  }>;
  [key: string]: any;
}

/**
 * Game session (subset needed for narrative)
 */
interface GameSession {
  flags: SessionFlags;
  [key: string]: any;
}

// ============================================================================
// ECS Component Access Helpers
// ============================================================================

/**
 * Ensure NPC components structure exists in session flags
 */
function ensureNpcComponents(session: GameSession, npcId: number): Record<string, any> {
  if (!session.flags.npcs) {
    session.flags.npcs = {};
  }

  const npcKey = Ref.npc(npcId);
  if (!session.flags.npcs[npcKey]) {
    session.flags.npcs[npcKey] = { components: {} };
  }

  if (!session.flags.npcs[npcKey].components) {
    session.flags.npcs[npcKey].components = {};
  }

  return session.flags.npcs[npcKey].components!;
}

/**
 * Get narrative runtime state for an NPC
 *
 * If no state exists, returns a fresh/empty state.
 */
export function getNarrativeState(
  session: GameSession,
  npcId: number
): NarrativeRuntimeState {
  const components = ensureNpcComponents(session, npcId);

  if (!components.narrative) {
    // Return fresh state
    return {
      activeProgramId: null,
      activeNodeId: null,
      stack: [],
      history: [],
      variables: {},
      lastStepAt: undefined,
      paused: false,
      error: undefined,
    };
  }

  return components.narrative as NarrativeRuntimeState;
}

/**
 * Set narrative runtime state for an NPC
 */
export function setNarrativeState(
  session: GameSession,
  npcId: number,
  state: NarrativeRuntimeState
): void {
  const components = ensureNpcComponents(session, npcId);
  components.narrative = state;
}

/**
 * Clear narrative runtime state for an NPC
 */
export function clearNarrativeState(
  session: GameSession,
  npcId: number
): void {
  const components = ensureNpcComponents(session, npcId);
  delete components.narrative;
}

// ============================================================================
// Program Lifecycle Helpers
// ============================================================================

/**
 * Start a new narrative program for an NPC
 *
 * If a program is already active, it will be pushed to the stack (nested program).
 */
export function startProgram(
  session: GameSession,
  npcId: number,
  programId: NarrativeProgramId,
  entryNodeId: NodeId,
  initialVariables?: Record<string, any>
): NarrativeRuntimeState {
  const state = getNarrativeState(session, npcId);

  // If there's an active program, push it to the stack
  if (state.activeProgramId && state.activeNodeId) {
    state.stack.push({
      programId: state.activeProgramId,
      nodeId: state.activeNodeId,
      pushedAt: Date.now(),
    });
  }

  // Set new active program
  state.activeProgramId = programId;
  state.activeNodeId = entryNodeId;
  state.variables = initialVariables || {};
  state.lastStepAt = Date.now();
  state.paused = false;
  state.error = undefined;

  // Add to history
  state.history.push({
    programId,
    nodeId: entryNodeId,
    timestamp: Date.now(),
  });

  // Save state
  setNarrativeState(session, npcId, state);

  return state;
}

/**
 * Finish the currently active program for an NPC
 *
 * If there are programs on the stack, pops and resumes the previous one.
 * Otherwise, clears the narrative state.
 */
export function finishProgram(
  session: GameSession,
  npcId: number
): NarrativeRuntimeState | null {
  const state = getNarrativeState(session, npcId);

  if (!state.activeProgramId) {
    // No active program
    return null;
  }

  // Pop from stack if available
  if (state.stack.length > 0) {
    const frame = state.stack.pop()!;
    state.activeProgramId = frame.programId;
    state.activeNodeId = frame.nodeId;
    state.lastStepAt = Date.now();

    // Add resume to history
    state.history.push({
      programId: frame.programId,
      nodeId: frame.nodeId,
      timestamp: Date.now(),
    });

    setNarrativeState(session, npcId, state);
    return state;
  } else {
    // No more programs, clear state
    clearNarrativeState(session, npcId);
    return null;
  }
}

/**
 * Advance to a new node in the current program
 */
export function advanceToNode(
  session: GameSession,
  npcId: number,
  nodeId: NodeId,
  choiceId?: string,
  edgeId?: string
): NarrativeRuntimeState {
  const state = getNarrativeState(session, npcId);

  if (!state.activeProgramId) {
    throw new Error('No active program to advance');
  }

  // Update active node
  state.activeNodeId = nodeId;
  state.lastStepAt = Date.now();

  // Add to history
  state.history.push({
    programId: state.activeProgramId,
    nodeId,
    timestamp: Date.now(),
    choiceId,
    edgeId,
  });

  // Save state
  setNarrativeState(session, npcId, state);

  return state;
}

/**
 * Pause the currently active program
 */
export function pauseProgram(
  session: GameSession,
  npcId: number
): NarrativeRuntimeState {
  const state = getNarrativeState(session, npcId);

  if (!state.activeProgramId) {
    throw new Error('No active program to pause');
  }

  state.paused = true;
  setNarrativeState(session, npcId, state);

  return state;
}

/**
 * Resume a paused program
 */
export function resumeProgram(
  session: GameSession,
  npcId: number
): NarrativeRuntimeState {
  const state = getNarrativeState(session, npcId);

  if (!state.activeProgramId) {
    throw new Error('No active program to resume');
  }

  state.paused = false;
  setNarrativeState(session, npcId, state);

  return state;
}

/**
 * Set error state for the current program
 */
export function setError(
  session: GameSession,
  npcId: number,
  errorMessage: string,
  nodeId: NodeId
): NarrativeRuntimeState {
  const state = getNarrativeState(session, npcId);

  state.error = {
    message: errorMessage,
    nodeId,
    timestamp: Date.now(),
  };

  setNarrativeState(session, npcId, state);

  return state;
}

/**
 * Clear error state
 */
export function clearError(
  session: GameSession,
  npcId: number
): NarrativeRuntimeState {
  const state = getNarrativeState(session, npcId);

  state.error = undefined;
  setNarrativeState(session, npcId, state);

  return state;
}

// ============================================================================
// Query Helpers
// ============================================================================

/**
 * Check if a program is currently active
 */
export function isProgramActive(
  session: GameSession,
  npcId: number,
  programId?: NarrativeProgramId
): boolean {
  const state = getNarrativeState(session, npcId);

  if (!state.activeProgramId) {
    return false;
  }

  if (programId) {
    return state.activeProgramId === programId;
  }

  return true;
}

/**
 * Get a program variable value
 */
export function getProgramVariable<T = any>(
  session: GameSession,
  npcId: number,
  variableName: string,
  defaultValue?: T
): T {
  const state = getNarrativeState(session, npcId);
  return state.variables[variableName] ?? defaultValue;
}

/**
 * Set a program variable value
 */
export function setProgramVariable(
  session: GameSession,
  npcId: number,
  variableName: string,
  value: any
): NarrativeRuntimeState {
  const state = getNarrativeState(session, npcId);
  state.variables[variableName] = value;
  setNarrativeState(session, npcId, state);
  return state;
}

/**
 * Check if a node has been visited in the history
 */
export function hasVisitedNode(
  session: GameSession,
  npcId: number,
  programId: NarrativeProgramId,
  nodeId: NodeId
): boolean {
  const state = getNarrativeState(session, npcId);

  for (const entry of state.history) {
    if (entry.programId === programId && entry.nodeId === nodeId) {
      return true;
    }
  }

  return false;
}

/**
 * Get the current call stack depth
 */
export function getStackDepth(
  session: GameSession,
  npcId: number
): number {
  const state = getNarrativeState(session, npcId);
  return state.stack.length;
}

/**
 * Get the currently active program ID and node ID
 */
export function getActiveProgram(
  session: GameSession,
  npcId: number
): { programId: NarrativeProgramId | null; nodeId: NodeId | null } {
  const state = getNarrativeState(session, npcId);
  return {
    programId: state.activeProgramId,
    nodeId: state.activeNodeId,
  };
}

/**
 * Get the execution history
 */
export function getHistory(
  session: GameSession,
  npcId: number
): NarrativeRuntimeState['history'] {
  const state = getNarrativeState(session, npcId);
  return state.history;
}

/**
 * Get the call stack
 */
export function getStack(
  session: GameSession,
  npcId: number
): NarrativeRuntimeState['stack'] {
  const state = getNarrativeState(session, npcId);
  return state.stack;
}
